library/src/test/java/bio/terra/landingzone/stairway/flight/create/reference/resource/step/BaseReferencedResourceStepTest.java

```diff
@@ -0,0 +1,287 @@
+package bio.terra.landingzone.stairway.flight.create.reference.resource.step;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.when;
+
+import bio.terra.landingzone.library.landingzones.definition.ArmManagers;
+import bio.terra.landingzone.library.landingzones.deployment.LandingZoneTagKeys;
+import bio.terra.landingzone.library.landingzones.deployment.ResourcePurpose;
+import bio.terra.landingzone.library.landingzones.deployment.SubnetResourcePurpose;
+import bio.terra.landingzone.service.landingzone.azure.model.LandingZoneRequest;
+import bio.terra.landingzone.stairway.common.model.TargetManagedResourceGroup;
+import bio.terra.landingzone.stairway.flight.LandingZoneFlightMapKeys;
+import bio.terra.landingzone.stairway.flight.create.resource.step.GetManagedResourceGroupInfo;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import com.azure.core.http.rest.PagedIterable;
+import com.azure.resourcemanager.AzureResourceManager;
+import com.azure.resourcemanager.resources.models.GenericResource;
+import com.azure.resourcemanager.resources.models.GenericResources;
+import com.azure.resourcemanager.resources.models.TagOperations;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+@Tag("unit")
+public class BaseReferencedResourceStepTest {
+
+  @Captor private ArgumentCaptor<Map<String, String>> tagsCaptor;
+
+  @Mock private TagOperations tagOperations;
+  @Mock private ArmManagers armManagers;
+  @Mock private AzureResourceManager azureResourceManager;
+  @Mock private GenericResources genericResources;
+  @Mock private GenericResource resource;
+  @Mock private PagedIterable<GenericResource> genericResourcesPage;
+
+  private final ArmResourceType armResourceType = ArmResourceType.STORAGE_ACCOUNT;
+  private final UUID landingZoneId = UUID.randomUUID();
+
+  @Mock private FlightContext context;
+  private FlightMap workingFlightMap;
+
+  @BeforeEach
+  void setUp() {
+    setUpMocks();
```

Comment:
A word of caution: this much shared setup can often be difficult to maintain, particularly when `STRICT_STUBS` is enabled (whcih it is, by default with `MockitoExtension`), which will fail if there are any mocks with uncalled stubs. 
A common effect of this is things like, if you change the order of a couple of operations, and you're testing the failure case of one of those operations, your test then fails because the mocks for the later operation isn't called. Then you end up having to change your setup for _everything_, often resulting in complex setup logic down the line.
I've found it's often better to repeat explicit setup for each test separately. It might be more verbose, but some repetition when writing tests is definitely ok.

I'm not saying that's necessarily an issue here; but it's something to think about.

Reply:
Yes, I agree. I had already refactored most of the mock setup, so it can be turned on/off as needed for the cases but left this initialization in place. Since I added a few more tests for idempotency, I refactored/removed the global mock setup. Now the cases wired up their mocks as the cases need.